firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,285 @@
+"""This module provides custom python preconditioners utilizing
+the Slate language.
+"""
+
+from __future__ import absolute_import, print_function, division
+
+import ufl
+
+from firedrake.matrix_free.preconditioners import PCBase
+from firedrake.petsc import PETSc
+from firedrake.slate.slate import Tensor
+
+
+__all__ = ['HybridizationPC']
+
+
+class HybridizationPC(PCBase):
+    """A Slate-based python preconditioner that solves a
+    mixed saddle-point problem using hybridization.
+
+    The forward eliminations and backwards reconstructions
+    are performed element-local using the Slate language.
+    """
+    def initialize(self, pc):
+        """Set up the problem context. Take the original
+        mixed problem and reformulate the problem as a
+        hybridized mixed system.
+
+        A KSP is created for the Lagrange multiplier system.
+        """
+        from ufl.algorithms.map_integrands import map_integrand_dags
+        from firedrake import (FunctionSpace, TrialFunction,
+                               TrialFunctions, TestFunction, Function,
+                               BrokenElement, MixedElement,
+                               FacetNormal, Constant, DirichletBC,
+                               Projector)
+        from firedrake.assemble import (allocate_matrix,
+                                        create_assembly_callable)
+        from firedrake.formmanipulation import ArgumentReplacer, split_form
+
+        # Extract the problem context
+        prefix = pc.getOptionsPrefix()
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+        test, trial = context.a.arguments()
+
+        V = test.function_space()
+        if V.mesh().cell_set._extruded:
+            # TODO: Merge FIAT branch to support TPC trace elements
+            raise NotImplementedError("Not implemented on extruded meshes.")
+
+        # Break the function spaces and define fully discontinuous spaces
+        broken_elements = [BrokenElement(Vi.ufl_element()) for Vi in V]
+        elem = MixedElement(broken_elements)
+        V_d = FunctionSpace(V.mesh(), elem)
+        arg_map = {test: TestFunction(V_d),
+                   trial: TrialFunction(V_d)}
+
+        # Replace the problems arguments with arguments defined
+        # on the new discontinuous spaces
+        replacer = ArgumentReplacer(arg_map)
+        new_form = map_integrand_dags(replacer, context.a)
+
+        # Create the space of approximate traces.
+        # The vector function space will have a non-empty value_shape
+        W = next(v for v in V if bool(v.ufl_element().value_shape()))
+        if W.ufl_element().family() in ["Raviart-Thomas", "RTCF"]:
+            tdegree = W.ufl_element().degree() - 1
+
+        else:
+            tdegree = W.ufl_element().degree()
```

Comment:
>Also, this only works for hybridising H(div) x L2 mixed problems, right? Or will it work if I use taylor-hood?

This PC assumes H(div) x L2. I doubt this will work for Taylor-Hood without modifications. This is because two sets of Lagrange multipliers will need to be introduced (I think) since we're breaking two continuous elements (P2-P1). This means the Schur-complement system will be mixed and we run into the "PyOP2 can't write into mixed dats" problem.

> what if the user is solving a Tensor-Valued x Vector-Valued problem? Is that even a legitimate thing?

I don't know. Can we think of such a formulation where we can apply hybridization? Perhaps some other form of static condensation, but that's not in the scope of this PC, I think. At least for this PR.